### ISSUE TITLE
Add file send/receive on Android

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#1809]
Tracking bug for sending/receiving files on Android.

**Sending:**

* add intent filter, e.g.
```xml
        <intent-filter>
            <action android:name="android.intent.action.SEND" />
            <category android:name="android.intent.category.DEFAULT"/>
            <data android:mimeType="application/*" />
            <data android:mimeType="audio/*" />
            <data android:mimeType="image/*" />
            <data android:mimeType="message/*" />
            <data android:mimeType="multipart/*" />
            <data android:mimeType="text/*" />
            <data android:mimeType="video/*" />
        </intent-filter>
        <intent-filter>
            <action android:name="android.intent.action.SEND_MULTIPLE" />
            <category android:name="android.intent.category.DEFAULT" />
            <data android:mimeType="application/*" />
            <data android:mimeType="audio/*" />
            <data android:mimeType="image/*" />
            <data android:mimeType="message/*" />
            <data android:mimeType="multipart/*" />
            <data android:mimeType="text/*" />
            <data android:mimeType="video/*" />
            <data android:mimeType="text/plain" />
        </intent-filter>
```

* upon starting that intent activity, choose a target device to share with (see `tailscale push --targets` implementation in `func runPushTargets` and how it calls `tailscale.FileTargets`): https://github.com/tailscale/tailscale/blob/3c9dea85e6a74e3a64d84323a25dbf8c42576099/cmd/tailscale/cli/push.go#L200

Then either do the HTTP PUT to the peer's peerapi URL directly (e.g. https://github.com/tailscale/tailscale/blob/3c9dea85e6a74e3a64d84323a25dbf8c42576099/cmd/tailscale/cli/push.go#L110) or if it's easier, using the localhost tailscale API ("localapi") handler to do it for you https://github.com/tailscale/tailscale/blob/3c9dea85e6a74e3a64d84323a25dbf8c42576099/ipn/localapi/localapi.go#L334 (the latter was added for iOS/macOS which live in a network sandbox and it was tedious for Swift code to do the sockopts to get out of the sandbox)

**Comment by bradfitz** [comment]
For recieving:

* listen for ipn.Notify messages with the new FilesWaiting bit (to say new files have fully arrived) and optionally IncomingFiles (for progress bars of incoming files): https://github.com/tailscale/tailscale/blob/3c9dea85e6a74e3a64d84323a25dbf8c42576099/ipn/backend.go#L74

Once a file is fully arrived; fetch it from the "inbox" (app-owned scratch dir) and copy it to the sdcard files/photos area with https://github.com/tailscale/tailscale/blob/3c9dea85e6a74e3a64d84323a25dbf8c42576099/cmd/tailscale/cli/debug.go#L105 (tailscale.WaitingFiles, tailscale.GetWaitingFile, tailscale.DeleteWaitingFile)

**Comment by bradfitz** [comment]
If you have any questions on UI, @rosszurowski can answer questions or share resources from other platforms.

**Comment by eliasnaur** [comment]
Thanks. @rosszurowski if you have iOS designs I'd like to see them so that our implementations won't needlessly differ.

**Comment by rosszurowski** [comment]
Yep! Best to get in sync up-front :)

Our general approach for this feature is using as much native UI as possible.

For sending files, we'll want to hook into Android's native share menu. It looks like apps get to open up a custom view with the selected media/files as an input. We'll leverage that system and introduce a custom view that looks something like the iOS view:

<img width="3454" alt="taildrop-ios" src="https://user-images.githubusercontent.com/303731/116312159-14b22500-a77a-11eb-8d4d-95a2d29220ca.png">

Notes on this design:
* We list all personal devices, online and offline that we know about. The thinking here is that we want to avoid shuffling this list around on users. If a device disappears or appears, the list shouldn't shuffle order and potentially change what you're tapping on. If a device transitions from online to offline, it should go offline, but its position shouldn't change. On iOS, I believe we implemented this view with a 5-second poll approach to updating online/offline status, but @soniaappasamy can give more context.
* The list of device is composed of the device name, device OS, connected status, and an icon representing the device. I have variations for phone, tablet, and desktop devices, but the specific heuristic we use here is up to you.
* Tapping a device will trigger a transfer. Tapping a row with an in-progress transfer will cancel that transfer. For most transfers of small files, it'll end before you can tap again, but that's okay, because cancelling is more relevant for large file transfers.
* I believe progress is calculated as the "bytes transferred / total bytes for all files being sent", but if you'd like to confirm what we do on iOS, @soniaappasamy can help.
* Upon completion of a file send, we change the progress circle into a checkmark, which persists for the duration of the share modal staying open. We foresee transferring to a single device being the most common case, but this design allows for the possibility to send to multiple devices. In the future, we may introduce a "sending" state that gives more details about each file being sent, but that's not in scope for the first version.
* Right now, we don't auto-close the share dialog on a successful transfer. This might change in the future, but the primary way to close the dialog is via the "Done" action in the top-right. On Android, it seems the more common pattern is an "X" in the top-left of the view, so let's go with that instead.
* For errors, we display a warning icon in place of the progress indicator, but we share specific language around the failure via a notification. Again, @soniaappasamy can give details the notification implementation (eg. how long it should persist for, what cases we need to include), but the general approach should minimize the amount of UI we have to build.

For receiving files, the ideal UI is… practically no UI. Since file transfers are restricted to your own devices for now (no inter-user file transfers), it's fine to automatically accept files and put them in the relevant places.

Here are our iOS designs:

<img width="1788" alt="ios-receive" src="https://user-images.githubusercontent.com/303731/116313943-72477100-a77c-11eb-9bb9-33777cc5c6c6.png">

Notes:
* We show a notification on file receipt. We'll want that on Android too. It'd be nice to leverage Android's ability to show progress in a notification to show that a transfer is in-progress and then notify when it's done. However, if that's difficult, we can do what iOS does and only notify on completion. Tapping the notification should open the file in it's end destination.
* An open question for Android: where should we send files? We'd like to integrate with the system as best possible. I'm happy to defer to you on what the best "native" location for different file types might be. If you'd rather not decide this, let me know and I can do some research. On iOS we have three categories: (1) photos/videos go to the camera roll, (2) URLs open directly in the browser, (3) and all other files go to the Files app. On macOS they land in ~/Downloads, and on iOS they land in On My iPhone/Tailscale/.
* Ideally, we'd show files directly in their destination as we're downloading them. On macOS, we initially had it appear in the downloads folder with a progress indicator. However, in the end we needed to fall back to a buffered approach where we download to a `tmp` folder and move to the destination once completed. If we need to buffer on Android too, that's perfectly okay :)
* The designs above show a white page with a spinner. That's because on iOS, the app needs to be in the foreground to take certain actions. Because Android doesn't have that limitation, we shouldn't need a page like that. However, if there are unforeseen complexities to building this on Android, please let me know — I'm happy to whip up some designs or jump on a  call to work through some options.

Phew. Let me know if you have questions or need more context. If there's any piece of this that introduces a lot of complexity, we can figure out a way to scope it down.

I'll let @bradfitz speak to timing, but we're not planning on including this in the v1.8 launch happening soon. This will likely be part of a follow-up patch release or v1.10, so there should be some time to work through the details.

**Comment by bradfitz** [comment]
Note that online/offline is via the new `tailcfg.Node.Online *bool` field. It's nil for devices whose online/offline you're not allowed to know. But it should generally be non-nil for your own devices (the only ones that are valid file sharing targets for now).

**Comment by soniaappasamy** [comment]
> On iOS, I believe we implemented this view with a 5-second poll approach to updating online/offline status, but @soniaappasamy can give more context.

We actually have not brought this in on iOS yet, but will probably do so in a 1.8.x release (unless it makes it in time for 1.8.0). The plan is to do a 5-second poll approach, like Ross said.

> * I believe progress is calculated as the "bytes transferred / total bytes for all files being sent", but if you'd like to confirm what we do on iOS, @soniaappasamy can help.

Yep, we show the "combined" progress of all of the selected files - so `total bytes transferred / total bytes to send`.

> Again, @soniaappasamy can give details the notification implementation (eg. how long it should persist for, what cases we need to include), but the general approach should minimize the amount of UI we have to build.

Yeah, on the iOS side, we show a notification in two cases currently: 1. when a send fails (we notify you that your files could not send) or 2. when receiving files. Both are just using the default iOS persistence length (they dismiss themselves, but are still accessible from the notification center).

Let me know if you need anymore context here!

**Comment by DentonGentry** [comment]
Is there anything more needed to support this for Android?

**Comment by eliasnaur** [comment]
Nope, the ball is in my court @DentonGentry.

However, now that I'm working on the feature, I'd like to get the device icons for the share dialog and the status icons (error, checkmark). Can you help me here, @rosszurowski? Thanks.

**Comment by eliasnaur** [comment]
I'm stuck. Connection attempts to the peerapi HTTP service seems to be dropped on Android if I connect from Tailscale peers. Local connections to the service does work. Help or pointers are greatly appreciated.

Details:

I'm working on the https://github.com/tailscale/tailscale-android/tree/file-sharing branch, which has a replace to a [debugging branch](https://github.com/eliasnaur/tailscale) of tailscale.

Netcat works locally:
```
$ adb shell
$  nc 100.86.189.59 45761
asd
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Connection: close

400 Bad Request
```

but not from my tailscaled Linux:

```
$ nc 100.86.189.59 45761
<hang>
```

Packages do arrive for remote connections:

```
05-31 18:43:25.645  8895  8916 I com.tailscale.ipn: 6.2M/148.6M magicsock: want call-me-maybe but endpoints stale; restunning
05-31 18:43:25.655  8895  8920 I com.tailscale.ipn: decode4: TCP{100.81.132.94:46808 > 100.86.189.59:45761} [69 0 0 60 48 99 64 0 64 6 0 24 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 232 142 0 0 2 4 4 216 4 2 8 10 251 176 33 237 0 0 0 0 1 3 3 7]
05-31 18:43:25.656  8895  8920 I com.tailscale.ipn: 6.4M/148.6M Accept: TCP{100.81.132.94:46808 > 100.86.189.59:45761} 60 tcp ok
05-31 18:43:25.656  8895  8920 I com.tailscale.ipn: multiTUN.runDevice: [4 0 0 0 174 166 9 79 0 0 0 0 0 0 0 0 69 0 0 60 48 99 64 0 64 6 0 24 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 232 142 0 0 2 4 4 216 4 2 8 10 251 176 33 237 0 0 0 0 1 3 3 7] err: <nil> written: 60
05-31 18:43:25.656  8895  8920 I com.tailscale.ipn: multiTun.Write: [4 0 0 0 174 166 9 79 0 0 0 0 0 0 0 0 69 0 0 60 48 99 64 0 64 6 0 24 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 232 142 0 0 2 4 4 216 4 2 8 10 251 176 33 237 0 0 0 0 1 3 3 7] err: <nil>
05-31 18:43:25.938  8895  8916 I com.tailscale.ipn: 6.6M/148.6M magicsock: STUN done; sending call-me-maybe to d:34f1c627d101af68 [/Vk0h]
05-31 18:43:26.684  8895  8919 I com.tailscale.ipn: decode4: TCP{100.81.132.94:46808 > 100.86.189.59:45761} [69 0 0 60 48 100 64 0 64 6 0 23 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 228 124 0 0 2 4 4 216 4 2 8 10 251 176 37 255 0 0 0 0 1 3 3 7]
05-31 18:43:26.684  8895  8919 I com.tailscale.ipn: 6.6M/148.6M Accept: TCP{100.81.132.94:46808 > 100.86.189.59:45761} 60 tcp ok
05-31 18:43:26.685  8895  8919 I com.tailscale.ipn: multiTUN.runDevice: [4 0 0 0 174 166 9 79 1 0 0 0 0 0 0 0 69 0 0 60 48 100 64 0 64 6 0 23 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 228 124 0 0 2 4 4 216 4 2 8 10 251 176 37 255 0 0 0 0 1 3 3 7] err: <nil> written: 60
05-31 18:43:26.685  8895  8919 I com.tailscale.ipn: multiTun.Write: [4 0 0 0 174 166 9 79 1 0 0 0 0 0 0 0 69 0 0 60 48 100 64 0 64 6 0 23 100 81 132 94 100 86 189 59 182 216 178 193 68 33 142 209 0 0 0 0 160 2 251 224 228 124 0 0 2 4 4 216 4 2 8 10 251 176 37 255 0 0 0 0 1 3 3 7] err: <nil>
```

but apparently they never reach the listener.

**Comment by eliasnaur** [comment]
Some progress. I still can't reach listeners created by the app process itself, but I can reach a separate Go program that listens on the Tailscale interface. I also tried using [VpnService.protect](https://developer.android.com/reference/android/net/VpnService#protect(int)) on the listening socket, to no avail.

**Comment by eliasnaur** [comment]
One workaround is to run the peerapi listener in a separate process. Unfortunately, every app process are covered by the same ip rules:

```
$ adb shell ip rule list
0:	from all lookup local
10000:	from all fwmark 0xc0000/0xd0000 lookup legacy_system
10500:	from all iif lo oif wlan0 uidrange 0-0 lookup wlan0
10500:	from all iif lo oif radio0 uidrange 0-0 lookup radio0
11000:	from all iif tun0 lookup local_network
12000:	from all fwmark 0xc0066/0xcffff lookup 1012
12000:	from all fwmark 0x0/0x20000 iif lo uidrange 0-10133 lookup tun0
12000:	from all fwmark 0x0/0x20000 iif lo uidrange 10135-99999 lookup tun0
12000:	from all fwmark 0xc0069/0xcffff lookup tun0
13000:	from all fwmark 0x10063/0x1ffff iif lo lookup local_network
13000:	from all fwmark 0x10065/0x1ffff iif lo lookup wlan0
13000:	from all fwmark 0x50064/0x5ffff iif lo lookup radio0
13000:	from all fwmark 0x10066/0x1ffff iif lo uidrange 0-0 lookup 1012
13000:	from all fwmark 0x10069/0x1ffff iif lo uidrange 0-10133 lookup tun0
13000:	from all fwmark 0x10069/0x1ffff iif lo uidrange 10135-99999 lookup tun0
13000:	from all fwmark 0x10069/0x1ffff iif lo uidrange 0-0 lookup tun0
14000:	from all iif lo oif wlan0 lookup wlan0
14000:	from all fwmark 0x40000/0x40000 iif lo oif radio0 lookup radio0
14000:	from all iif lo oif tun0 uidrange 0-10133 lookup tun0
14000:	from all iif lo oif tun0 uidrange 10135-99999 lookup tun0
15000:	from all fwmark 0x0/0x10000 lookup legacy_system
16000:	from all fwmark 0x0/0x10000 lookup legacy_network
17000:	from all fwmark 0x0/0x10000 lookup local_network
19000:	from all fwmark 0x65/0x1ffff iif lo lookup wlan0
21000:	from all fwmark 0x69/0xffff lookup wlan0
22000:	from all fwmark 0x0/0xffff iif lo lookup wlan0
32000:	from all unreachable
```

I confirmed that by running the separate test listener through `adb shell run-as com.tailscale.ipn`.

**Comment by eliasnaur** [comment]
@bradfitz, @DentonGentry I'd appreciate it if you could take a look. The blocking issue seems to be rooted in a low-level networking configuration problem, and you have a more experience dealing with those than I.

**Comment by eliasnaur** [comment]
Ok, connection accept packets (SYN-ACK?) are routed through the underlying network, not the tunnel device:

```
$ adb root
$ adb shell
# ip route get 100.89.247.103     // Tailscale IP of the host.
100.89.247.103 dev tun0 table tun0 src 100.80.71.11 uid 0.     // Routing through the tun0 interface.
    cache
# run-as com.tailscale.ipn    // run as the Tailscale app user (10134).
$ ip route get 100.89.247.103
100.89.247.103 via 10.0.2.2 dev wlan0 table 1003 src 10.0.2.16 uid 10134.   // Routing through the wlan0 interface.
    cache
```

I've confirmed that with tcpdump.

Curiously, connection resets (RST?) are delivered through the tunnel as expected. I can confirm that by `nc`'ing an unbound port.

**Comment by Maltahl** [comment]
@eliasnaur could the workaround be to add a static route for each tailscale device address as a /32 and point tailscale as nexthop ? Or is that out of the question ?

**Comment by eliasnaur** [comment]
I'm not sure what you mean. The routing table while Tailscale is active is:

```
# ip route show table all
default via 10.0.2.2 dev wlan0 table wlan0 proto static
10.0.2.0/24 dev wlan0 table wlan0 proto static scope link
default via 10.0.2.2 dev radio0 table radio0 proto static
10.0.2.0/24 dev radio0 table radio0 proto static scope link
100.80.71.11 dev tun0 table tun0 proto static scope link
100.81.132.94 dev tun0 table tun0 proto static scope link
100.89.247.103 dev tun0 table tun0 proto static scope link
100.94.141.97 dev tun0 table tun0 proto static scope link
100.99.151.70 dev tun0 table tun0 proto static scope link
100.100.100.100 dev tun0 table tun0 proto static scope link
100.101.102.103 dev tun0 table tun0 proto static scope link
10.0.2.0/24 dev radio0 proto kernel scope link src 10.0.2.15
10.0.2.0/24 dev wlan0 proto kernel scope link src 10.0.2.16
broadcast 10.0.2.0 dev radio0 table local proto kernel scope link src 10.0.2.15
broadcast 10.0.2.0 dev wlan0 table local proto kernel scope link src 10.0.2.16
local 10.0.2.15 dev radio0 table local proto kernel scope host src 10.0.2.15
local 10.0.2.16 dev wlan0 table local proto kernel scope host src 10.0.2.16
broadcast 10.0.2.255 dev radio0 table local proto kernel scope link src 10.0.2.15
broadcast 10.0.2.255 dev wlan0 table local proto kernel scope link src 10.0.2.16
local 100.80.71.11 dev tun0 table local proto kernel scope host src 100.80.71.11
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1
fe80::/64 dev wlan0 table wlan0 proto kernel metric 256 pref medium
fe80::/64 dev wlan0 table wlan0 proto static metric 1024 pref medium
fec0::/64 dev wlan0 table wlan0 proto kernel metric 256 expires 86122sec pref medium
fec0::/64 dev wlan0 table wlan0 proto static metric 1024 pref medium
default via fe80::2 dev wlan0 table wlan0 proto ra metric 1024 expires 1522sec hoplimit 64 pref medium
fe80::/64 dev radio0 table radio0 proto static metric 1024 pref medium
fec0::/64 dev radio0 table radio0 proto kernel metric 256 expires 86122sec pref medium
fec0::/64 dev radio0 table radio0 proto static metric 1024 pref medium
default via fe80::2 dev radio0 table radio0 proto ra metric 1024 expires 1522sec hoplimit 64 pref medium
fd7a:115c:a1e0:ab12:4843:cd96:6250:470b dev tun0 table tun0 proto kernel metric 256 pref medium
fd7a:115c:a1e0:ab12:4843:cd96:6250:470b dev tun0 table tun0 proto static metric 1024 pref medium
fd7a:115c:a1e0::/48 dev tun0 table tun0 proto static metric 1024 pref medium
fe80::/64 dev tun0 table tun0 proto kernel metric 256 pref medium
fe80::/64 dev radio0 proto kernel metric 256 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
local fd7a:115c:a1e0:ab12:4843:cd96:6250:470b dev tun0 table local proto kernel metric 0 pref medium
local fe80::15:b2ff:fe00:0 dev wlan0 table local proto kernel metric 0 pref medium
local fe80::454e:6ae2:870a:cda6 dev tun0 table local proto kernel metric 0 pref medium
local fe80::98cc:7dff:fee2:d06c dev radio0 table local proto kernel metric 0 pref medium
local fec0::15:b2ff:fe00:0 dev wlan0 table local proto kernel metric 0 pref medium
local fec0::25f2:28:bd1b:9c1e dev wlan0 table local proto kernel metric 0 pref medium
local fec0::98cc:7dff:fee2:d06c dev radio0 table local proto kernel metric 0 pref medium
local fec0::fcbf:a7be:d417:aab1 dev radio0 table local proto kernel metric 0 pref medium
ff00::/8 dev radio0 table local metric 256 pref medium
ff00::/8 dev wlan0 table local metric 256 pref medium
ff00::/8 dev tun0 table local metric 256 pref medium
```

The device local ipv4 address is 100.80.71.11 and that appears to be set up correctly. Note also that the problem is that packets outgoing from the device is sent to wlan0, not through tun0. Incoming packets are routed to tun0 as expected.

In case I missed something obvious, I added a call to VPNService.Builder.addRoute for every address we add through VPNService.Builder.addAddress. There was no change in behaviour.

**Comment by bradfitz** [comment]
@eliasnaur, we're all actively fighting this now too. Related bug is #2102.

**Comment by bradfitz** [comment]
@eliasnaur, I just pushed https://github.com/tailscale/tailscale-android/commit/ac8ec020b847a1d3e68b859f3daa521cf629ebe1 to update Android to Tailscale 1.10 and sent out PR https://github.com/tailscale/tailscale-android/pull/12 for review. That should unblock you on this bug.

**Comment by Maltahl** [comment]
The update broke tailscale on my phone. GUI wont load like on #1965 
OnePlus 8T running Android 11

**Comment by danackerson** [comment]
> The update broke tailscale on my phone. GUI wont load like on #1965
> OnePlus 8T running Android 11

Ditto on Pixel 4 XL running Android 11.

**Comment by DentonGentry** [comment]
We'll look at the issue with the 1.10.1 release in https://github.com/tailscale/tailscale/issues/2290.

I had built an APK locally containing the file sharing code https://github.com/tailscale/tailscale-android/pull/14 and didn't run into the issue, presumably the problem is something not directly related to that.
